src/marqo/tensor_search/models/add_docs_objects.py

```diff
@@ -1,17 +1,137 @@
+        """Ensure that weights is provided if and only if the type is 'MappingsObjectType.multimodal_combination'."""
+        is_multimodal = values.get('type', None) == enums.MappingsObjectType.multimodal_combination
+        if is_multimodal and weights is None:
+            raise errors.InvalidArgError(
+                f"Error validating multimodal combination object. Reason: 'weights' is not provided"
+                f"\nRead about the mappings object here: https://docs.marqo.ai/0.0.15/API-Reference/mappings/"
+            )
+        elif not is_multimodal and weights is not None:
+            raise errors.InvalidArgError(
+                f"Error validating mappings. Mapping is not multimodal combination object. Reason: field is not multimodal, but weights provided."
+                f"\nRead about the mappings object here: https://docs.marqo.ai/0.0.15/API-Reference/mappings/"
+            )
+        return weights
+
+class Document(BaseModel):
+    _id: Optional[str] = None
+
+    class Config:
+        extra = 'allow'
+
+    # @validator('*', pre=True, allow_reuse=True)
+    def validate_field_names(self, v, values, field):
+        if field.name == "_id":
+            return
+
+        validation.validate_field_name(field)
+
+    # @root_validator(allow_reuse=True)
+    def validate_id(cls, values):
+        if "_id" in values.keys():
+            validation.validate_id(values.get("_id"))
+
+        return values
 
+    # @root_validator(allow_reuse=True)
+    def validate_document_size(cls, values):
+        """Ensure document is less than MARQO_MAX_DOC_BYTES bytes (when set)."""
 
-@dataclass(frozen=True, config=AddDocsParamsConfig)
-class AddDocsParams:
+        max_doc_size = utils.read_env_vars_and_defaults(var=enums.EnvVars.MARQO_MAX_DOC_BYTES)
+        if max_doc_size is None:    
+            return
+
+        try:
+            serialized = json.dumps(values)
+        except TypeError as e:
+            raise errors.InvalidArgError(
+                f"Unable to index document: it is not serializable! Document: `{values}` "
+            )
+
+        if len(serialized) > int(max_doc_size):
+            maybe_id = f" _id:`{values['_id']}`" if '_id' in values else ''
+            raise errors.DocTooLargeError(
+                f"Document{maybe_id} with length `{len(serialized)}` exceeds "
+                f"the allowed document size limit of [{max_doc_size}]."
+            )
+    
+        return values
+
+    def _validate_field_name(self, key: str):
+        if key == "_id":
+            return 
+        
+        validation.validate_field_name(key)
+
+    def validate_document(self) -> None:
+        """
+        Validate the document object.
+        """
+        self.validate_document_size(self.dict())
+        self.validate_id(self.dict())
+        [self._validate_field_name(k) for (k, _) in self.dict().items()]
+
+    def get_or_generate_id(self, id_gen: Callable[[None], str] = lambda: str(uuid.uuid4())) -> str:
```

Comment:
We can add tests to it. I don't really see why it needs a separate function and file. The Id is a product of a document and should be associated with the data structure. It might be worth distinguishing between Document and DocumentWithID but in the current flow because they come in interchangeably, you'd have a need to split `List[Union[Document, DocumentWithID]]`

Reply:
Worth adding some tests too

Reply:
Added explicit unit tests for this